src/server/services/procedures/salio/salio.js

```diff
@@ -0,0 +1,749 @@
+ */
+SalIO.prototype._getRegistered = function () {
+    var state = this._state,
+        sensors = [];
+    for (var mac_addr in state.registered) {
+        if (this._sensors[mac_addr].isMostlyAlive()) {
+            sensors.push(mac_addr);
+        } else {
+            delete state.registered[mac_addr];
+        }
+    }
+    return sensors;
+};
+
+// /**
+//  * Registers for receiving messages from the given sensors.
+//  * @param {array} sensors one or a list of sensors
+//  * @deprecated
+//  */
+// SalIO.prototype.eavesdrop = function (sensors) {
+//     return this.listen(sensors);
+// };
+
+// /**
+//  * Registers for receiving messages from the given sensors.
+//  * @param {array} sensors one or a list of sensors
+//  */
+// SalIO.prototype.listen = async function (sensors) {
+//     var state = this._state;
+
+//     for (var mac_addr in state.registered) {
+//         if (this._sensors[mac_addr]) {
+//             this._sensors[mac_addr].removeClientSocket(this.socket);
+//         }
+//     }
+//     state.registered = {};
+
+//     if (!Array.isArray(sensors)) {
+//         sensors = ('' + sensors).split(/[, ]/);
+//     }
+
+//     var ok = true;
+//     for (var i = 0; i < sensors.length; i++) {
+//         var sensor = await this._getSensor(sensors[i]);
+//         if (sensor) {
+//             state.registered[sensor.mac_addr] = sensor;
+//             sensor.addClientSocket(this.socket);
+//         } else {
+//             ok = false;
+//         }
+//     }
+//     return ok;
+// };
+
+/**
+ * Returns the addresses of all authorized sensors.
+ * @returns {array}
+ */
+SalIO.prototype.getSensors = async function () {
+    const availableSensors = Object.keys(this._sensors);
+    let sensors = await acl.authorizedRobots(this.caller.username, availableSensors);
+    return sensors;
+};
+
+/**
+ * Performs the pre-checks and maps the incoming call to a sensor action.
+ * @param {String} fnName name of the method/function to call on the sensor object
+ * @param {Array} args array of arguments
+ */
+SalIO.prototype._passToSensor = async function (fnName, args) {
+    args = Array.from(args);
+    let sensorId = args.shift();
+    const sensor = await this._getSensor(sensorId);
+    if (sensor && sensor.accepts(this.caller.clientId)) {
+        let rv = sensor[fnName](sensor, args);
+        if (rv === undefined) rv = true;
+        return rv;
+    }
+    return false;
+};
+
+/**
+ * Get the color code for the specified red green and blue levels.
+ * @param {BoundedNumber<0, 255>} red red level (0-255).
+ * @param {BoundedNumber<0, 255>} green green level (0-255).
+ * @param {BoundedNumber<0, 255>} blue blue level (0-255).
+ * @returns {boolean} The orientation angles relative to the Earth's magnetic field.
+ */
+SalIO.prototype.getColor = function (red, green, blue) {
+    return 0xff000000 | ((red & 0xff) << 16) | ((green & 0xff) << 8) | (blue & 0xff);
+};
+
+if (SALIO_MODE === 'native' || SALIO_MODE === 'both') {
```

Comment:
Do we want to have all this extra code that looks like it was copied from RoboScape?

Reply:
That part I'm not sure. So far I've kept it in case we want to use it for a cyber security lesson later on (maybe the dangers of people hacking a GPS-capable networked device).

Reply:
Gotcha. Let's not worry about it for now but keep it in mind going forward. I actually have some concerns about the scalability of handling it this way for RoboScape as well but we can handle these concerns later. (It seems we are handling a class-level config parameter on the entire deployment.)